operator/apis/operator/v1alpha1/oapserverconfig_types.go

```diff
@@ -0,0 +1,76 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// OAPServerConfigSpec defines the desired state of OAPServerConfig
+type OAPServerConfigSpec struct {
+	// Version of OAP.
+	//+kubebuilder:validation:Required
+	Version string `json:"version,omitempty"`
+	// StaticConfig holds the OAP server static configuration.
+	// +kubebuilder:validation:Optional
+	StaticConfig []corev1.EnvVar `json:"staticConfig,omitempty"`
+	// DynamicConfig holds the OAP server dynamic configuration.
+	// +kubebuilder:validation:Optional
+	DynamicConfig map[string]string `json:"dynamicConfig,omitempty"`
```

Comment:
The purpose of my design is to minimize the coupling with the oapserver config. When the oapserver configuration in the main repo changes, there is no need to modify it here, what do you think? BTW，`splitting it into a series` is to deal each [config key](https://skywalking.apache.org/docs/main/latest/en/setup/backend/dynamic-config/#single-configuration) as a field of CR? @hanahmily

Reply:
No, you don't have to dive into the details of each file. the CRD should support transforming CRD to the [group config](https://skywalking.apache.org/docs/main/latest/en/setup/backend/dynamic-config-configmap/#group-config), for example:

```yaml
apiVersion: operator.skywalking.apache.org/v1alpha1
kind: OAPServerDynamicConfig
metadata:
  name: core.default.endpoint-name-grouping-openapi
spec:
  customerAPI-v1: value of customerAPI-v1
  productAPI-v1: value of productAPI-v1
```

You could extract the group key from each item by leveraging the CRD to make the whole file neater.

Reply:
Make sense to me, in my view, we should support the [group config](https://skywalking.apache.org/docs/main/latest/en/setup/backend/dynamic-config-configmap/#group-config) and the 
 [single config](https://skywalking.apache.org/docs/main/latest/en/setup/backend/dynamic-config/#single-configuration). We can use an `array` to store the data rather than a `map`.

Reply:
Yep, you had better support both of them. Not all config-values are YAML. A single config CRD is good in this case. 

> We should use an array to store the data rather than a map.

Why does it port to an array? Would you elaborate it a bit here?

Reply:
Emm, what  I thought before is Using a "config-value" `array` can support more configuration in spec, and we don't need to set the value like `customerAPI-v1` as a field of the CRD. For example, we can support the following CRs.

```yaml
apiVersion: operator.skywalking.apache.org/v1alpha1
kind: OAPServerDynamicConfig
metadata:
  name: core.default.endpoint-name-grouping-openapi.v1
spec:
  - config: core.default.endpoint-name-grouping-openapi.customerAPI-v1
     value: value of customerAPI-v1
  - config: core.default.endpoint-name-grouping-openapi.productAPI-v1
     value: value of productAPI-v1
```

```yaml
apiVersion: operator.skywalking.apache.org/v1alpha1
kind: OAPServerDynamicConfig
metadata:
  name: agent-analyzer.default.slowDBAccessThreshold
spec:
  - config: agent-analyzer.default.slowDBAccessThreshold
     value: default:200,mongodb:50
```

```yaml
apiVersion: operator.skywalking.apache.org/v1alpha1
kind: OAPServerDynamicConfig
metadata:
  name: Skywalking-Dynamic-Config
spec:
  - config: core.default.endpoint-name-grouping-openapi.customerAPI-v1
     value: value of customerAPI-v1
  - config: core.default.endpoint-name-grouping-openapi.productAPI-v1
     value: value of productAPI-v1
  - config: agent-analyzer.default.slowDBAccessThreshold
     value: default:200,mongodb:50
  - config: core.default.apdexThreshold
     value : |-
       default: 500
       # example:
       # the threshold of service "tomcat" is 1s
       # tomcat: 1000
       # the threshold of service "springboot1" is 50ms
       # springboot1: 50
```

So users can put a single configuration in a CR for brevity or put all configurations in a CR for convenience. BTW, 
it seems that we can also use `map`, like the following.

```yaml
apiVersion: operator.skywalking.apache.org/v1alpha1
kind: OAPServerDynamicConfig
metadata:
  name: core.default.endpoint-name-grouping-openapi.v1
spec:
   config:
      core.default.endpoint-name-grouping-openapi.customerAPI-v1: value of customerAPI-v1
      core.default.endpoint-name-grouping-openapi.productAPI-v1: value of productAPI-v1
```

```yaml
apiVersion: operator.skywalking.apache.org/v1alpha1
kind: OAPServerDynamicConfig
metadata:
  name: agent-analyzer.default.slowDBAccessThreshold
spec:
  config:
     agent-analyzer.default.slowDBAccessThreshold: default:200,mongodb:50
```

What do you think about it? Is there a better way or something I missed here?

Reply:
`Array` style is good for me, here is my some suggestions:

- You had better replace `config` with `name`. `config` is the combination of `name` and `value`. 
- As I recommended, you have to remove the global prefix from each item.
- Add root node `data` in the `spec`.

```yaml
apiVersion: operator.skywalking.apache.org/v1alpha1
kind: OAPServerDynamicConfig
metadata:
  name: core.default.endpoint-name-grouping-openapi
spec:
  data:
  - name: customerAPI-v1
     value: value of customerAPI-v1
  - name: productAPI-v1
     value: value of productAPI-v1
```

For those who have only a single item, you could leave an empty `name` as below:

```yaml
apiVersion: operator.skywalking.apache.org/v1alpha1
kind: OAPServerDynamicConfig
metadata:
  name: agent-analyzer.default.slowDBAccessThreshold
spec:
  data:
  -  value: default:200,mongodb:50
```

An all in one version should like 

```yaml
apiVersion: operator.skywalking.apache.org/v1alpha1
kind: OAPServerDynamicConfig
metadata:
  name: Skywalking-Dynamic-Config
spec:
  data:
      - name: core.default.endpoint-name-grouping-openapi.customerAPI-v1
         value: value of customerAPI-v1
      - name: core.default.endpoint-name-grouping-openapi.productAPI-v1
         value: value of productAPI-v1
      - name: agent-analyzer.default.slowDBAccessThreshold
         value: default:200,mongodb:50
      - name: core.default.apdexThreshold
         value : |-
           default: 500
           # example:
           # the threshold of service "tomcat" is 1s
           # tomcat: 1000
           # the threshold of service "springboot1" is 50ms
           # springboot1: 50
````

Reply:
Thanks for the detailed explain, I understand it. As for static files lay in [resources](https://github.com/apache/skywalking/tree/master/oap-server/server-starter/src/main/resources), I think we can use the same way, the `metadata.name` could be their path and the `spec.data.name` could be the specific file.

Reply:
It that possible we use [`RawExtension`](https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime#RawExtension) for value?

It is much more flexible. The official `kube-scheduler` uses it for configuration.

Reply:
Thanks for your suggestion, I find an [example](https://stackoverflow.com/questions/66637260/parse-byte-array-for-yaml-value-kubebuilder) and the code is [here](https://go.dev/play/p/syx8v7gAmDH). 

In my understanding, the [RawExtension](https://kubernetes.io/zh/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#rawextension) can only store the object defined in k8s cluster. 

But the configuration is not defined in k8s cluster. If we use the `RawExtension`, it looks like we need to define an object for each config? I think a string can hold the value simply, what do you think?

Reply:
> In my understanding, the [RawExtension](https://kubernetes.io/zh/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#rawextension) can only store the object defined in k8s cluster.

No. As long as you mark the field with `// +kubebuilder:pruning:PreserveUnknownFields`, api-server will preserve everything you store in this field.

The `RawExtenstion` is actually serialized to `json` format, tools like `kubectl` will implicitly transform it to `yaml`. So we will see something like,

```yaml
apiVersion: operator.skywalking.apache.org/v1alpha1
kind: OAPServerDynamicConfig
metadata:
  name: Skywalking-Dynamic-Config
spec:
  data:
      - name: core.default.endpoint-name-grouping-openapi.customerAPI-v1
         value: value of customerAPI-v1
      - name: core.default.endpoint-name-grouping-openapi.productAPI-v1
         value: value of productAPI-v1
      - name: agent-analyzer.default.slowDBAccessThreshold
         value: default:200,mongodb:50
      - name: core.default.apdexThreshold
         value:
           default: 500 
```

Reply:
@lujiajing1126, make sense to me, I know your meaning. Using `RawExtenstion` can make it easier for users to modify the configuration just like `YAML` rather than `String`. As @hanahmily said, currently, when we create a CR, the fields in it will simply be fulfilled in the configmap.  Later, if we want to verify the configuration, I think it is better to use `RawExtenstion`.  What do you think?

Reply:
Since these values are directly mapped to `configmap`s, I think string may be enough now.